File: Java/Homework06_04.java
/*
 * File name : Homework06_04.java
 * Author : Kim Yeong Ho
 * Date : July 18, 2023
 * Major feature : 
 * - 다음은 2차원 상의 한 점을 표현하는 Point 클래스이다.
 * - Point를 상속받아 양수의 공간에서만 점을 나타내는 PositivePoint 클래스를 작성하라.
 * - 다음 main() 메소드를 포함하고 실행 결과와 같이 출력되게 하라.
 * - (10, 10)의 점입니다.
 * - (10, 10)의 점입니다.
 * - (0, 0)의 점입니다.
 */
class Point {
	private int x, y;
	public Point(int x, int y) { this.x = x; this.y = y; }
	public int getX() { return x; }
	public int getY() { return y; }
	protected void move(int x, int y) { this.x = x; this.y = y; }
}

class PositivePoint extends Point {
	public PositivePoint() { super(0, 0); }
	public PositivePoint(int x, int y) { super(0, 0); if (x > 0 && y > 0) super.move(x, y); }
	public void move(int x, int y) { super.move(getX(), getY()); if (x > 0 && y > 0) super.move(x, y); }
	public String toString() { return ("("+ getX() + ", " + getY() + ")의 점"); }
}

public class Homework06_04 {
	public static void main(String[] args) {
		PositivePoint p = new PositivePoint();
		p.move(10, 10);
		System.out.println(p.toString() + "입니다.");
		
		p.move(-5, 5);
		System.out.println(p.toString() + "입니다.");
		
		PositivePoint p2 = new PositivePoint(-10, -10);
		System.out.println(p2.toString() + "입니다.");
	}
}

File: C/Homework06_04/Homework06_04.c
/*
* File name : Homework06_04.c
* Author : Kim Yeong Ho
* Date : July 28, 2023
* Major feature :
* - 메뉴 선택에 따라 동작하는 간이 스택을 구현하라
*   - 배열 주소 이외에는 배열을 직접 참조할 수 없음
*   - 각 함수의 데이터타입은 void 이어야 함
* - 구현 필수 함수
*   - void pop(int *a)
*   - void push(int *a)
*   - void dump(int *a)
*/
#include <stdio.h>

#define ARRAY_SIZE	100

int top;
void print_menu();
void init(int* a);
void push(int* a);
void pop(int* a);
void dump(int* a);

int main(void) {
	int input, arr[ARRAY_SIZE];

	init(arr);

	while (1) {
		print_menu();
		scanf("%d", &input);

		switch (input) {
			case 1: push(arr);							break;
			case 2: pop(arr);							break;
			case 3: dump(arr);							break;
			case 4:	return 0;							break;
			default: printf("다시 입력해 주십시오.\n");	break;
		}
		printf("\n");
	}

	return 0;
}

void print_menu() {
	printf("1. 데이터 입력받기\n");
	printf("2. 데이터 빼기\n");
	printf("3. 전체 데이터 출력하기\n");
	printf("4. 종료하기\n");
	printf("메뉴선택 : ");
}

void init(int* a) {
	top = 0;
	for (int i = 0; i < ARRAY_SIZE; i++)
		*a++ = NULL;
}

void push(int* a) {
	printf("스택에 넣을 데이터를 입력하세요 : ");
	scanf("%d", a + top);
	top++;
}

void pop(int* a) {
	printf("pop 된 데이터 : %d\n", *(a + top));
	*(a + top) = NULL;
	top--;
}

void dump(int* a) {
	printf("dump data = ");
	for (int i = 0; i < top; i++) {
		printf("%d ", *(a + i));
	}
	printf("\n");
}
